apps/www/_blog/2022-12-16-pg-graphql-v1.mdx

```diff
@@ -0,0 +1,100 @@
+
+### ACID - Atomicity, Consistency, Isolation, and Durability
+
+Databases have strong ACID guarantees. Being embedded in the database lets us claim those guarantees through the GraphQL API. For example, if any part of a multi-mutation GraphQL request fails, the entire request can roll back to leave the database in a consistent state.
+
+
+## From SQL to Rust 
+
+Our pure [SQL implementation of the pg_graphql transpiler](https://github.com/supabase/pg_graphql/blob/bd0283718abaf329d98c69808f862594e9df5edc/pg_graphql--0.4.0.sql) carried us from v0.0.1 to v0.4.0. Ultimately, we started to feel some pain, although probably not in the way you’d expect.
+
+Opinions about business logic in SQL are notoriously split. In my view, there are some foot guns, but:
+
+- SQL has types
+- SQL has functions
+- Functions are composable
+
+Ergo, keep your functions pure and you’ve got a strongly typed functional programming paradigm. Taking that approach, the developer experience for pg_graphql was surprisingly excellent. The feedback cycle with [pg_regress](https://www.postgresql.org/docs/9.1/regress.html) is fast, and occasionally the type system even caught a bug or two. What more could you ask for?
+
+Somewhere around the 4,000 line count, things started to bog down. The rough shape of the codebase was coming into view, which made it clear where we could further extract common patterns into functions. At this point I learned two things:
+
+- SQL functions are “pass by value” (copy) so passing state around can get expensive
+- The query planner tends to stop inlining functions when calls are deeply nested
+
+Combine those two facts with our GraphQL AST being a big ball of JSONB, and the GitHub issues about slow schema introspection on free tier started to trickle in.
+
+Enter [pgx](https://github.com/tcdi/pgx)… 
+
+pgx is a rust framework for building Postgres extensions. It has a very polished onboarding experience and (as you would expect) full SQL interop. That made it straightforward to initially sprinkle some rust on pg_graphql’s hotspots. When that went well, we ultimately rewrote pg_graphql from the ground up as a pure Rust project for the v0.5.0 release. 
+
+One of rust’s main selling points is zero cost abstractions. That means high-level concepts like generics incur no runtime penalty. Transitioning from SQL’s expensive abstractions to rust’s zero cost abstractions has been hugely satisfying as it enabled refactoring the codebase into a more maintainable state. Development velocity is up. Code quality is up. Performance is WAY up.
+
+As of 1.0, overhead introduced by pg_graphql is sub 300 *micro*seconds per request on free-tier hardware. When executed from SQL we see~1060 queries per second per connection (no parallelism). If we include the entire Auth + HTTP stack, free tier can handle ~645 requests per second.
```

Comment:
```suggestion
As of 1.0, overhead introduced by pg_graphql is sub 300 *micro*seconds per request on free-tier hardware. When executed from SQL we see ~1060 queries per second per connection (no parallelism). If we include the entire Auth + HTTP stack, free tier can handle ~645 requests per second.
```

I'd add the "On larger instance the stack handles upwards of 10k requests / second." up into this paragraph too